corehq/apps/api/tests/test_case_update_api.py

```diff
@@ -0,0 +1,179 @@
+
+        case = self.case_accessor.get_case(case.case_id)
+        self.assertEqual(case.name, 'Beth Harmon')
+        self.assertEqual(case.owner_id, 'us_chess_federation')
+        self.assertEqual(case.dynamic_case_properties(), {
+            'champion': 'true',
+            'dob': '1948-11-02',
+            'rank': '2100',
+            'sport': 'chess',
+        })
+
+    def test_create_child_case(self):
+        parent_case = self._make_case()
+        res = self.post('create_case', {
+            '@case_type': 'match',
+            '@case_name': 'Harmon/Luchenko',
+            '@owner_id': 'harmon',
+            'properties': {
+                'external_id': '23',
+                'winner': 'Harmon',
+            },
+            'indices': {
+                'parent': {
+                    '@case_id': parent_case.case_id,
+                    '@case_type': 'player',
+                    '@relationship': 'child',
+                },
+            },
+        })
+        self.assertEqual(res.keys(), ['@case_id', '@form_id'])
+
+        case = self.case_accessor.get_case(res['@case_id'])
+        self.assertEqual(case.name, 'Harmon/Luchenko')
+        self.assertEqual(case.owner_id, 'harmon')
+        self.assertEqual(case.dynamic_case_properties(), {
+            'external_id': '23',
+            'winner': 'Harmon',
+        })
+        self.assertEqual(case.indices[0].identifier, 'parent')
+        self.assertEqual(case.indices[0].referenced_id, parent_case.case_id)
+        self.assertEqual(case.indices[0].referenced_type, 'player')
+        self.assertEqual(case.indices[0].relationship_id, 'child')
+        self.assertEqual(case.indices[0].case.case_id, parent_case.case_id)
+
+    def test_bulk_action(self):
+        existing_case = self._make_case()
+        res = self.post('bulk_update', [
+            {
+                # update existing case
+                '@case_id': existing_case.case_id,
+                '@case_name': 'Beth Harmon',
+                '@owner_id': 'us_chess_federation',
+                'properties': {
+                    'rank': '2100',
+                    'champion': 'true',
+                },
+            },
+            {
+                # No case_id means this is a new case
+                '@case_type': 'player',
+                '@case_name': 'Jolene',
+                '@owner_id': 'methuen_home',
+                'properties': {
+                    'external_id': '2',
+                    'sport': 'squash',
+                    'dob': '1947-03-09',
+                },
+            },
+        ])
+        #  only returns a single form ID - chunking should happen in the client
+        self.assertEqual(res.keys(), ['@case_ids', '@form_id'])
+
+        updated_case = self.case_accessor.get_case(existing_case.case_id)
+        self.assertEqual(updated_case.name, 'Beth Harmon')
+
+        new_case = self.case_accessor.get_case(res['@case_ids'][1])
```

Comment:
I don't love this.  I want the bulk API's return value to tell the client what case IDs were assigned in some fashion.  Here it assumes that the list of returned IDs is positionally aligned with the list of cases sent in.

Another option would be to return a map of external IDs to created case IDs, but external IDs aren't required.  A further option would be to allow providing your own case IDs (even if only in the bulk API), which would obviate the problem, and allow creating cases with relationships in a single request.

Reply:
Great points.  I was envisioning a relatively low limit for the number of updates allowed in a single request (like 1000, maybe even 100, like the case importer), and commit them in a single form, forcing the client to chunk bulk actions on their side.  With that in mind, I think it should be fine to process synchronously and return a complete status.

I was picturing success/failure being monolithic - if there are any validation issues with any case in a request, the whole request fails, and we'd return a 400.

> What if we returned the whole case as it now exists on the server for these requests?

That would avoid that issue.  You don't think that's overkill?  Followup question there - we could return the serialization from the regular case API, but that uses different property names and a different format (`case_name`, `case_type`, and `date_opened` are inside `properties` and not prefixed with `@`).

Reply:
> I was picturing success/failure being monolithic - if there are any validation issues with any case in a request, the whole request fails, and we'd return a 400.

I suppose this is the only way it could work since we are assuming a single form submission / transaction per request. 

I don't think it is overkill to return the whole case, unless you think that it would be weirdly resource intensive - although if you are limiting things to 1000 cases/request this seems reasonable. I think if you bumped the API version, you wouldn't have the serialization issue.